blob/conformance.go

```diff
@@ -0,0 +1,460 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package blob provides an easy way to interact with Blob objects within
+// a bucket. It utilizes standard io packages to handle reads and writes.
+package blob
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"io/ioutil"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+// bucketMaker describes functions used to create a bucket for a test.
+// Multiple calls to bucketMaker during a test run must refer to the same
+// storage bucket.
+// Functions should return the bucket along with a "done" function to be called
+// when the test is complete.
+// If bucket creation fails, functions should t.Fatal to stop the test.
+type bucketMaker func(t *testing.T) (*Bucket, func())
```

Comment:
Export this type (otherwise godoc won't show up).

Also, if this should be returning the same bucket every time, why not just have the caller handle cleanup and pass a `*Bucket` directly to the conformance test?

Reply:
Rename: done.

Same bucket: They must refer to the same storage bucket, but the *Bucket doesn't have to be the same. For example, for "same storage bucket", I had to change the fileblob implementation to use a static subdirectory under /tmp, instead of a dynamic one each time. For "*Bucket doesn't have to be the same", the GCS/AWS implementations may use a different HTTP recorder/replayer for each call (based on t.Name()).

These are used, for example, by testRead, which creates a blob once and then runs a bunch of read tests over it. It calls MakeBucket once at the beginning for the creation, and then repeatedly after that for each sub-test of read. 
-- If the storage bucket was different every time, the "read" sub-tests would fail because they wouldn't see the blob created at the beginning.
-- Each sub-test has a separate HTTP replay file (plus there's a separate one for the blob creation/deletion).